packages/ra-core/src/dataProvider/useDataProvider.ts

```diff
@@ -0,0 +1,294 @@
+import { useCallback, useContext } from 'react';
+import { Dispatch } from 'redux';
+import { useDispatch, useSelector } from 'react-redux';
+
+import DataProviderContext from './DataProviderContext';
+import validateResponseFormat from './validateResponseFormat';
+import undoableEventEmitter from './undoableEventEmitter';
+import {
+    startOptimisticMode,
+    stopOptimisticMode,
+} from '../actions/undoActions';
+import { FETCH_END, FETCH_ERROR, FETCH_START } from '../actions/fetchActions';
+import { showNotification } from '../actions/notificationActions';
+import { refreshView } from '../actions/uiActions';
+import { ReduxState, DataProvider } from '../types';
+
+type DataProviderHookFunction = (
+    type: string,
+    resource: string,
+    params: any,
+    options?: UseDataProviderOptions
+) => Promise<{ data?: any; total?: any; error?: any }>;
+
+interface UseDataProviderOptions {
+    action?: string;
+    meta?: object;
+    undoable?: boolean;
+    onSuccess?: any;
+    onFailure?: any;
+}
+
+const defaultDataProvider = () => Promise.resolve(); // avoids adding a context in tests
+
+/**
+ * Hook for getting an instance of the dataProvider as prop
+ *
+ * Gets a dataProvider function, which behaves just like the real dataProvider
+ * (same signature, returns a Promise), but dispatches Redux actions along the
+ * process. The benefit is that react-admin tracks the loading state when using
+ * this function, shows the loader animation while the dataProvider is waiting
+ * for a response, and executes side effects when the response arrives.
+ *
+ * In addition to the 3 parameters of the dataProvider function (verb, resource,
+ * payload), the returned function accepts a fourth parameter, an object
+ * literal which may contain side effects, or make the action optimistic (with
+ * undoable: true).
+ *
+ * @return dataProvider (type, resource, payload, options) => Promise<any>
+ *
+ * @example
+ *
+ * import React, { useState } from 'react';
+ * import { useDispatch } from 'react-redux';
+ * import { useDataProvider, showNotification } from 'react-admin';
+ *
+ * const PostList = () => {
+ *      const [posts, setPosts] = useState([])
+ *      const dispatch = useDispatch();
+ *      const dataProvider = useDataProvider();
+ *
+ *      useEffect(() => {
+ *          dataProvider('GET_LIST', 'posts', { filter: { status: 'pending' }})
+ *            .then(({ data }) => setPosts(data))
+ *            .catch(error => dispatch(showNotification(error.message, 'error')));
+ *      }, [])
+ *
+ *      return (
+ *          <Fragment>
+ *              {posts.map((post, key) => <PostDetail post={post} key={key} />)}
+ *          </Fragment>
+ *     }
+ * }
+ */
+const useDataProvider = (): DataProviderHookFunction => {
+    const dispatch = useDispatch() as Dispatch;
+    const dataProvider = useContext(DataProviderContext) || defaultDataProvider;
+    const isOptimistic = useSelector(
```

Comment:
I'm not sure we need this reducer to handle optimistic detection. Can't w achieve the same with metas on the action ?

Reply:
No, because the use case is the following:

1. user clicks on Delete post
2. the `useDataProvider` hook optimistically deletes the record from the store and redirects to the post list
3. the list calls the dataProvider for the list of posts on mount
4. While waiting for the result, the list shows the posts from the sorte, where the deleted post has been removed
5. As the deletion wasn't performed on the server yet, the server returns results including the deleted record
6. The list shows the deleted post again

To avoid that, the 'optimistic mode' blocks queries while an optimistic update is pending, so 3 and 5 never happen.